### PR TITLE
I2c support

### PR DESCRIPTION
**Pull request by m0n5t3r**
Related to #12 - I decided to split it in two; this one adds I2C support and separates the communication concerns. It also adds an untested safeload implementation for ADAU145x (it's simple enough, the only thing I'm unsure about is the format of the number of words - I _think_ it's correct, because the compiler output example they show has the address and the number as int32, and this is not something used in calculations, but I've been surprised by sigmastudio in my brief experience with it).

The safeload method is not used at this stage, it will come with the next PR (ADAU1701 support and `sigma_tcp_server` changes).

### CODE REVIEW

**Comment by m0n5t3r** [comment]
allright, done with it :)

**Comment by elagil** [comment]
Thanks, I will have a look!

**Review by elagil — COMMENTED** [review]
Here comes a first batch of comments! Thanks again for your contribution so far.

File: sigmadsp/hardware/i2c.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+        logger.info("Starting I2C handling thread.")
+        self.thread = threading.Thread(target=self.serve_forever, name="I2C handler thread")
+        self.thread.daemon = True
+        self.thread.start()
+
+    def _initialize_i2c(self, bus: int = 1, device: int = 0x34):
+        """Initialize the I2C hardware.
+
+        Bus will be 1 for RaspberryPi hardware.
```
[the remark this answers is not in the document]
Comment (m0n5t3r) [reply sigmadsp/hardware/i2c.py:38]: https://pinout.xyz/ seems to suggest it's stable across versions; it's definitely not true for other boards of the same format (a banana pi pro I have has I2C2 there, for instance)

File: sigmadsp/hardware/adau14xx.py [hunk 2]
```diff
@@ -137,3 +143,21 @@ def adjust_volume(self, adjustment_db: float, address: int) -> float:
         )
 
         return linear_to_db(new_volume)
+
+    def safeload(self, address: int, data: bytes, count: int = 1):
```
[the remark this answers is not in the document]
Comment (m0n5t3r) [reply sigmadsp/hardware/adau14xx.py:147]: ah, so it does have weird surprises then... I have no means of testing this, but https://www.analog.com/media/en/technical-documentation/data-sheets/ADAU1452_1451_1450.pdf doesn't seem to mention it (the software features chapter, page 93):

> The address_SafeLoad parameter is the target address in parameter RAM. This designates the first address to be written in the safeload transfer. If more than one word is written, the address increments automatically for each data-word. 
>
> The num_SafeLoad parameter designates the number of words to be written. For a biquad filter algorithm, the number of words to be written is five because there are five coefficients in a biquad IIR filter. For a simple mono gain algorithm, the number of words to be written is one. This parameter also serves as the trigger; when it is written, a safeload write is triggered on the next frame.

File: sigmadsp/hardware/i2c.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+from smbus2 import SMBus, i2c_msg
+
+from sigmadsp.helper.conversion import int16_to_bytes
+
+# A logger for this module
+logger = logging.getLogger(__name__)
+
+
+class I2cHandler:
```
[the remark this answers is not in the document]
Comment (m0n5t3r) [reply sigmadsp/hardware/i2c.py:14]: sounds good to me

[the remark this answers is not in the document]
Comment (m0n5t3r) [reply sigmadsp/hardware/i2c.py]: done, tested I2C and it still works; I don't see why SPI wouldn't work, unless I made some typo

**Comment by m0n5t3r** [comment]
addressed everything, I think; I'll check for replies in the morning :)

File: sigmadsp/application/frontend.py [hunk 4]
```diff
@@ -87,7 +87,7 @@ def main():
     if arguments.address is not None:
         backend_address = arguments.address
 
-    response: ControlResponse
+    response: ControlResponse = None
```
Comment (elagil) [inline sigmadsp/application/frontend.py:90]: I wonder why mypy doesn't complain, I would have expected this to be required:
```suggestion
    response: Union[ControlResponse, None] = None
```

File: sigmadsp/hardware/adau14xx.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (elagil) [reply sigmadsp/hardware/adau14xx.py:147]: Ok, I am not sure. I have to read more about the topic.

File: sigmadsp/hardware/i2c.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (elagil) [reply sigmadsp/hardware/i2c.py]: Ok, I will try to make a suggestion tomorrow or so.

**Comment by m0n5t3r** [comment]
I already did the splitting in cade4b17d2d68d3005ad6e3670b3eb6471a43ee3 btw

**Comment by m0n5t3r** [comment]
sorry to be pushy, but I'm not sure how you merge things (many people are very like to destroy merged branch history via squashing; if you're one of them, it means I can't start a new branch from the current one, submit the second PR for ADAU1701 / TCPIP1701 protocol support and forget about this until I get a review / change request, and it's bugging me to have uncommitted code)

**Comment by elagil** [comment]
I will not squash, only rebase and merge. You can start a branch on this pull request and rebase onto mainafter this is merged.

File: sigmadsp/hardware/dsp.py [hunk 5]
```diff
@@ -101,6 +109,13 @@ def parse_config(self):
         except (KeyError, TypeError):
             logger.info("No DSP pin definitions were found in the configuration file.")
 
+        self.type = self.config["dsp"]["type"]
+        self.protocol = self.config["dsp"].get("protocol", "spi")
```
Comment (elagil) [inline sigmadsp/hardware/dsp.py:113]: I would suggest to add type hints for these two fields in the `__init__` function, just for readability.

Regarding the default (`spi`), I prefer to throw an error and let the user correct their config file. The protocol should not be an optional parameter.

File: sigmadsp/hardware/dsp.py [hunk 6]
```diff
@@ -101,6 +109,13 @@ def parse_config(self):
...
+
+        if self.protocol == "i2c":
+            self.i2c_address = self.config["dsp"].get("i2c_address", 0x38)
+            self.i2c_bus = self.config["dsp"].get("i2c_bus", 1)
```
Comment (elagil) [inline sigmadsp/hardware/dsp.py:117]: Please add some log message for when a default value is chosen (so, no value found in config).

**Review by elagil — COMMENTED** [review]
Here is a batch of (minor) comments, but overall it already looks very good.

File: sigmadsp/hardware/base.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+from multiprocessing import Pipe
+
+# A logger for this module
+logger = logging.getLogger(__name__)
+
+
+class HandlerBase:
+    """Base class for communication handlers talking to SigmaDSP chipsets."""
+
+    # cosmetic use only
```
[the remark this answers is not in the document]
Comment (m0n5t3r) [reply sigmadsp/hardware/base.py:14]: It didn't cross my mind to use `__class__.__name__`, I'll remove the field

File: sigmadsp/hardware/dsp.py (hunk 5 above)
Comment (m0n5t3r) [reply sigmadsp/hardware/dsp.py:113]: I went for the thing that would break the fewest installations :)

going to add errors and make configuration options more uniform (`bus_number` and `device_address` instead of `i2c_bus`, `i2c_address`, `spi_bus`, `spi_address`)

File: sigmadsp/hardware/dsp.py (hunk 6 above)
Comment (m0n5t3r) [reply sigmadsp/hardware/dsp.py:117]: hm, I think it might be better to actually require specifying the bus / address in the configuration

File: sigmadsp/hardware/dsp.py (hunk 5 above)
Comment (elagil) [reply sigmadsp/hardware/dsp.py]: You are right, I didn't even think about existing installations. In that case, this will be a breaking release, so it is more obvious.

**Comment by m0n5t3r** [comment]
hm, pre-commit worked fine on my end...

File: sigmadsp/hardware/dsp.py (hunk 6 above)
Comment (elagil) [reply sigmadsp/hardware/dsp.py]: Yes, I agree.

**Comment by elagil** [comment]
> hm, pre-commit worked fine on my end...

Are you using a newer Python version? I think from 3.10 onwards, types like `dict` can be used instead of `typing.Dict`.

**Comment by m0n5t3r** [comment]
raspbian seems to use 3.9.2

File: sigmadsp/hardware/base.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    def write(self, address: int, data: bytes):
+        """Write data over the hardware interface via the pipe to the hardware thread.
+
+        Args:
+            address (int): DSP register address to write to
+            data (bytes): Binary data to write
+        """
+        self.pipe_end_owner.send("write")
+        self.pipe_end_owner.send((address, data))
```
[the remark this answers is not in the document]
Comment (elagil) [reply sigmadsp/hardware/base.py]: See this new issue https://github.com/elagil/sigmadsp/issues/16

File: sigmadsp/hardware/base.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+        Args:
+            address (int): DSP register address to read from
+            length (int): Number of bytes to read
+
+        Returns:
+            bytes: Register content
+        """
+        self.pipe_end_owner.send("read")
+        self.pipe_end_owner.send((address, length))
```
[the remark this answers is not in the document]
Comment (elagil) [reply sigmadsp/hardware/base.py]: Again, see https://github.com/elagil/sigmadsp/issues/16.

File: sigmadsp/hardware/adau14xx.py [hunk 10]
```diff
@@ -137,3 +142,22 @@ def adjust_volume(self, adjustment_db: float, address: int) -> float:
...
+        """Write data to the chip using software safeload.
+
+        Args:
+            address (int): Address to write to
+            data (bytes): Data to write; multiple parameters should be concatenated
+            count (int): Number of 4 byte words to write
+        """
+        count = min(count, 5)
```
Comment (elagil) [inline sigmadsp/hardware/adau14xx.py:154]:
```suggestion
        count = min(count, len(Adau14xx.SAFELOAD_DATA_REGISTERS))
```

However, shouldn't we actually raise an exception, if we wanted to write more than is possible with the safeload registers? E.g.:

```suggestion
        if count > len(Adau14xx.SAFELOAD_DATA_REGISTERS):
            raise IndexError(f"Cannot write {count * 4} bytes by means of software safeload, the maximum is {len(Adau14xx.SAFELOAD_DATA_REGISTERS) * 4} bytes.")
```

File: sigmadsp/hardware/adau14xx.py [hunk 11]
```diff
@@ -137,3 +143,22 @@ def adjust_volume(self, adjustment_db: float, address: int) -> float:
...
+        for register_index, register_address in zip(range(count), Adau14xx.SAFELOAD_DATA_REGISTERS):
+            data_buf = data[
+                register_index * self.FIXPOINT_REGISTER_LENGTH : (register_index + 1) * self.FIXPOINT_REGISTER_LENGTH
+            ]
+            self.write(register_address, data_buf)
```
[the remark this answers is not in the document]
Comment (elagil) [reply sigmadsp/hardware/adau14xx.py:159]: Tested and working, but I only tested for `count == 1`

[the remark this answers is not in the document]
Comment (m0n5t3r) [reply sigmadsp/hardware/adau14xx.py:159]: I guess the only issue that could arise is data passed from sigmastudio (if you can convince it to update multiple parameters at the same time - I'm not sure if the code that deals with software safeload will do this, the one that does the hardware safeload on ADAU1701 will only safeload one value at a time); for setting your own parameters, it's only a matter of formatting the data as a bytes buffer, which is the responsibility of the calling code, which you control.

TL;DR, I think it's safe to assume it works and deal with any sigmastudio related issues in the sigmastudio protocol implementation.

[the remark this answers is not in the document]
Comment (elagil) [reply sigmadsp/hardware/adau14xx.py:163]: Sounds reasonable!

**Comment by elagil** [comment]
Thanks for your work!
